core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/STDCMGraph.kt

```diff
@@ -98,36 +99,71 @@ class STDCMGraph(
+        val maxMarginDuration = estimateMaxMarginDuration(node)
+        val baseNodeCost = node.timeSinceDeparture + node.remainingTimeEstimation
+        val visitedNodesParameters =
+            VisitedNodes.Parameters(
+                null,
+                node.time,
+                node.maximumAddedDelay,
+                maxMarginDuration,
+                baseNodeCost
+            )
         if (node.locationOnEdge != null) {
             val explorer = node.infraExplorer.clone()
-            val fingerprint =
+            visitedNodesParameters.fingerprint =
                 VisitedNodes.Fingerprint(
                     explorer.getLastEdgeIdentifier(),
                     node.waypointIndex,
                     node.locationOnEdge.distance
                 )
-            if (visitedNodes.isVisited(fingerprint, node.time)) return listOf()
-            visitedNodes.markAsVisited(fingerprint, node.time)
+            if (visitedNodes.isVisited(visitedNodesParameters)) return listOf()
+            visitedNodes.markAsVisited(visitedNodesParameters)
             res.addAll(STDCMEdgeBuilder.fromNode(this, node, explorer).makeAllEdges())
         } else {
             val extended = extendLookaheadUntil(node.infraExplorer.clone(), 3)
             for (newPath in extended) {
                 if (newPath.getLookahead().size == 0) continue
                 newPath.moveForward()
-                val fingerprint =
+                visitedNodesParameters.fingerprint =
                     VisitedNodes.Fingerprint(
                         newPath.getLastEdgeIdentifier(),
                         node.waypointIndex,
                         0.meters
                     )
-                if (visitedNodes.isVisited(fingerprint, node.time)) return listOf()
-                visitedNodes.markAsVisited(fingerprint, node.time)
+                if (visitedNodes.isVisited(visitedNodesParameters)) return listOf()
+                visitedNodes.markAsVisited(visitedNodesParameters)
                 res.addAll(STDCMEdgeBuilder.fromNode(this, node, newPath).makeAllEdges())
             }
         }
         return res
     }
 
+    /**
+     * Give a (rough) estimation of how much delay we could add before this node with engineering
+     * margins. Should be on the pessimistic side.
+     */
+    private fun estimateMaxMarginDuration(inputNode: STDCMNode): Double {
+        // We look for the 20km before the node (very rough estimation of a distance that lets the
+        // train slow down to a stop and speed up). We return the max delay that can be added after
+        // the train in all of those edges, on top of the start delay
+        var node = inputNode
+        var remainingDistance = 20_000.meters
+        var maxTime = Double.POSITIVE_INFINITY
+        while (true) {
+            val edge = node.previousEdge ?: return 0.0
+
+            val endTime = edge.timeStart + edge.totalTime + edge.maximumAddedDelayAfter
```

Comment:
It doesn't change the math, since we substract the following values, but i feel like it'd be clearer: the real start time is `edge.timeStart + edge.addedDelay`. Same with `edge.timeNextOccupancy + edge.addedDelay`.